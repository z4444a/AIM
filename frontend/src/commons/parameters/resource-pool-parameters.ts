import ResourcePoolModel from '../../model/get/resource-pool-model';
import { PageAndSortParams } from '../../model/parameters/page-and-sort-params';

export default interface ResourcePoolParameters extends PageAndSortParams<ResourcePoolModel> {
  name?: string;
  type?: number;
  active?: boolean | undefined;
  onlyMine?: boolean | undefined;
}
