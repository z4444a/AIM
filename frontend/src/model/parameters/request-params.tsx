import RequestModel from '../get/request-model';
import { PageAndSortParams } from './page-and-sort-params';

export interface RequestParams extends PageAndSortParams<RequestModel> {
  typeName?: string | number;
  description?: string;
  statusId?: string | number;
}
