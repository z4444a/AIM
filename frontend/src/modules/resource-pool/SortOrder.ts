import { Order } from './Order';

export type SortOrder = Order.ASC | Order.DESC | undefined;
