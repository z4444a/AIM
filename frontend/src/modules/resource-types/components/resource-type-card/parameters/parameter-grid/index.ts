import { withTranslation } from 'react-i18next';
import { ParameterGrid } from './parameter-grid';

export default withTranslation('common')(ParameterGrid);
