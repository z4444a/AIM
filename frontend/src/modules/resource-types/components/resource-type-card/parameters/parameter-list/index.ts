import { ParameterList } from './parameter-list';
import { withTranslation } from 'react-i18next';
import withStyles from '@material-ui/core/styles/withStyles';
import { styles } from './styles';

export default withTranslation('common')(withStyles(styles)(ParameterList));
