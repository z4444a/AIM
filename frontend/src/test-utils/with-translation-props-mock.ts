import { WithTranslation } from 'react-i18next';

export function getMockWithTranslationProps(): WithTranslation {
  // eslint-disable-next-line
  const t = ((key: string) => `${key}_i18n`) as any;
  // eslint-disable-next-line
  const i18n = {} as any;
  const props: WithTranslation = {
    tReady: true,
    t,
    i18n,
  };

  return props;
}
