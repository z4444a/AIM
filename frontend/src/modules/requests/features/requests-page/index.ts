import { RequestParams } from '../../../../model/parameters/request-params';
import {
  fetchAllRequests,
  setRequestsPageFilterParam,
} from '../../../../redux/actions/request-actions';
import { push } from 'connected-react-router';
import { connect } from 'react-redux';
import { RequestsPage } from './requests-page';
import { RootState } from 'typesafe-actions';
import { withTranslation } from 'react-i18next';
import { toggleRequestPageFilterPanel } from '../../../../redux/actions/portal-actions';
import {
  fetchResourceTypesSuggestions,
  resetResourceTypeParameters,
  setResourceTypeParameters,
  updateRequestFilters,
} from '../../../../redux/actions/request-filter-panel-actions';
import { ParameterModifier } from '../../../../model/parameter-modifier';
import FullParameterModel from '../../../../model/get/full-parameter-model';
import { CustomFilterCmpContext } from '../../../../commons/components/aim-data-grid/aim-data-grid';

export const mapStateToProps = (state: RootState) => ({
  requests: state.requests.page,
  typeSuggestions: state.requestFilterPanel.resourceTypeSuggestion.suggestions,
  params: state.requests.params,
  filters: state.requestFilterPanel.filters,
  showFilterPanel: state.portal.requestPageFilterPanelOpen,
  resourceTypeParametersList: state.requestFilterPanel.resourceTypeParameters
    ? state.requestFilterPanel.resourceTypeParameters.parameters.filter(
        value => value.modifier === ParameterModifier.REQUEST_PARAMETER
      )
    : [],
  user: state.tokens !== null ? state.tokens.user : null,
});

export const mapDispatchToProps = (dispatch: Function) => ({
  fetchRequests: (params?: RequestParams) => dispatch(fetchAllRequests.request(params)),
  fetchTypeSuggestions: () => dispatch(fetchResourceTypesSuggestions.request()),
  setResourceTypeParameters: (parameters: FullParameterModel[]) =>
    dispatch(setResourceTypeParameters(parameters)),
  resetResourceTypeParameters: () => dispatch(resetResourceTypeParameters()),
  redirect: (path: string) => dispatch(push(path)),
  toggleFilterPanel: (showFilterPanel: boolean) =>
    dispatch(toggleRequestPageFilterPanel(showFilterPanel)),
  updateFilters: (filters: CustomFilterCmpContext) => dispatch(updateRequestFilters(filters)),
  resetParams: () => dispatch(setRequestsPageFilterParam({ pageSize: 10 })),
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(withTranslation('common')(RequestsPage));
