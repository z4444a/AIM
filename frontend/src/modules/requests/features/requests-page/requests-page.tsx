import React from 'react';
import { Path } from '../../../../commons/path';
import {
  FilterValueType,
  Props as DataGridProps,
} from '../../../../commons/components/aim-data-grid/aim-data-grid';
import AimDataGrid from '../../../../commons/components/aim-data-grid';
import { SortOrder } from '../../../resource-pool/SortOrder';
import { MUIDataTableColumn } from 'mui-datatables';
import { WithTranslation } from 'react-i18next';
import { mapDispatchToProps, mapStateToProps } from './index';
import moment from 'moment';
import { DateFormat } from '../../../../commons/values/date-format';
import { RequestParams } from '../../../../model/parameters/request-params';
import RequestModel from '../../../../model/get/request-model';
import NamedModel from '../../../../model/base/named-model';
import { RequestState } from '../../../../model/get/request-state';
import AddCircle from '@material-ui/icons/AddCircle';
import FiberManualRecord from '@material-ui/icons/FiberManualRecord';
import IconButton from '@material-ui/core/IconButton';
import Tooltip from '@material-ui/core/Tooltip';
import RequestFilterPanel from '../../components/request-filter-panel/index';
import { FilterField } from '../../components/request-filter-panel/request-filter-panel';
import RequestStatus from '../../components/request-status/index';
import { SuggestionItem } from '../../../../commons/components/aim-autocomplete/aim-autocomplete';
import FullParameterModel from '../../../../model/get/full-parameter-model';
import { ParameterType } from '../../../../model/parameter-type';
import {
  DateFieldValue,
  GenericFieldValue,
  GenericFormValue,
  ListFieldValue,
  NumberFieldValue,
  TextFieldValue,
} from '../../../../commons/components/generic-fields/models/field-config.model';
import ParamValueShortVer from '../../../../model/get/param-value-short-ver';
import { Role } from '../../../../commons/role';
import { ParameterModifier } from '../../../../model/parameter-modifier';
import { parameterShortVersionList } from '../../../../commons/components/parameter-short-version-list/parameter-short-version-list';

export type Props = ReturnType<typeof mapDispatchToProps> &
  ReturnType<typeof mapStateToProps> &
  WithTranslation;

export function getFieldValue(
  param: FullParameterModel,
  field: GenericFormValue
): GenericFieldValue | undefined {
  const item: GenericFormValue = field ? field : {};
  const paramValues = item[param.id];
  const type = param.parameterType ? param.parameterType : '';
  switch (type) {
    case ParameterType.DATE: //TODO: multiple
      return (paramValues[0] as unknown) as DateFieldValue;
    case ParameterType.NUMBER:
      return (paramValues[0] as unknown) as NumberFieldValue;
    case ParameterType.REAL:
      return (paramValues[0] as unknown) as NumberFieldValue;
    case ParameterType.TEXT:
      return paramValues[0] as TextFieldValue;
    case ParameterType.LIST:
      const param = (paramValues[0] as unknown) as ListFieldValue;
      return param ? param[0] : undefined;
    default:
      return undefined;
  }
}

export class RequestsPage extends React.Component<Props> {
  private readonly colDefs: MUIDataTableColumn[];
  private readonly gridConfig: Partial<DataGridProps>;

  public constructor(props: Props) {
    super(props);
    this.gridConfig = this.getGridConfig();
    this.colDefs = this.getGridColumnsConfig();
  }

  public componentDidMount(): void {
    const { fetchTypeSuggestions } = this.props;
    fetchTypeSuggestions();
    this.props.fetchRequests(this.props.params);
  }

  public componentWillUnmount(): void {
    const { resetParams } = this.props;
    resetParams();
  }

  public render(): JSX.Element {
    const { requests, params, showFilterPanel } = this.props;
    const sort = {
      sortBy: params.sortBy as string,
      direction: params.order,
    };

    if (!requests) {
      return <div />;
    }

    const filter: { [key: string]: FilterValueType } = {};
    if (params.typeName) {
      filter[FilterField.RESOURCE_TYPE] = params.typeName;
    }
    if (params.description) {
      filter[FilterField.DESCRIPTION] = params.description;
    }
    if (params.statusId) {
      filter[FilterField.STATUS] = params.statusId;
    }
    return (
      <AimDataGrid
        columns={this.colDefs}
        data={requests.content}
        totalPageCount={requests.totalElements}
        rowsPerPage={params.pageSize}
        sort={sort}
        filterPanelOpen={showFilterPanel}
        filterValues={filter}
        {...this.gridConfig}
      />
    );
  }

  private getGridConfig(): Partial<DataGridProps> {
    const { t } = this.props;
    const gridConfig: Partial<DataGridProps> = {
      pageTitle: t('requestPage.header'),
      onColumnSortChange: (changedColumn, direction) =>
        this.handleColumnSortChange(changedColumn, direction),
      onChangePage: currentPage => this.handleChangePage(currentPage),
      onChangeRowsPerPage: numberOfRows => this.handleChangeRowsPerPage(numberOfRows),
      renderCustomFilterCmp: () => (
        <RequestFilterPanel
          context={this.getValue()}
          getResourceTypeSuggestions={this.getResourceTypeSuggestions}
          onResourceTypeSelected={this.handleResourceTypeSelected}
          resourceTypeParameters={this.getParameters()}
          onFiltersChanged={this.handleOnFiltersChanged}
        />
      ),
      onFilterChange: this.handleFilterChange,
      toggleFilterPanel: this.toggleShowFilterPanel,
      customToolbar: () => {
        return (
          <Tooltip title={t('requestPage.buttons.createRequest')}>
            <IconButton onClick={() => this.props.redirect(Path.REQUEST_CREATE)}>
              <AddCircle />
            </IconButton>
          </Tooltip>
        );
      },
      onRowClick: (rowData, rowMeta, dataItem) => this.handleRedirect(dataItem as RequestModel),
    };

    return gridConfig;
  }

  private toggleShowFilterPanel = (showFilterPanel: boolean) => {
    const { toggleFilterPanel } = this.props;
    toggleFilterPanel(showFilterPanel);
  };

  private getValue = (): { [key: string]: FilterValueType } => {
    return this.props.filters.values ? this.props.filters.values : {};
  };

  private getParameters = (): FullParameterModel[] => {
    return this.props.filters &&
      this.props.filters.values &&
      this.props.filters.values[FilterField.RESOURCE_TYPE]
      ? this.props.resourceTypeParametersList
      : [];
  };

  private handleOnFiltersChanged = (value: { [key: string]: FilterValueType }) => {
    const { updateFilters, filters } = this.props;
    const filter = Object.assign({}, filters);
    filter.values = value;
    updateFilters(filter);
  };

  private getResourceTypeSuggestions = (value: string) => {
    const { typeSuggestions } = this.props;
    const result: SuggestionItem[] =
      !typeSuggestions || typeSuggestions.length === 0
        ? []
        : typeSuggestions.map(item => {
            const suggestionItem: SuggestionItem = {
              value: item.name,
              key: item.id,
            };
            return suggestionItem;
          });
    return result.filter(item =>
      item.value.toLocaleLowerCase().includes(value.toLocaleLowerCase())
    );
  };

  private getGridColumnsConfig(): MUIDataTableColumn[] {
    const { t, user } = this.props;
    const colDefs: MUIDataTableColumn[] = [
      {
        label: t('requestPage.grid.columns.id'),
        name: 'id',
        options: {
          display: 'false',
        },
      },
      {
        label: t('requestPage.grid.columns.creation'),
        name: 'creation',
        options: {
          customBodyRender: (creation: number) => moment(creation).format(DateFormat.ISO_DATE_TIME),
        },
      },
      {
        label: t('requestPage.grid.columns.type'),
        name: 'type',
        options: {
          customBodyRender: (type: NamedModel) => (type ? type.name : ''),
        },
      },
      {
        label: t('requestPage.grid.columns.project'),
        name: 'project',
        options: {
          customBodyRender: (project: NamedModel) => (project ? project.name : ''),
        },
      },
      {
        label: t('requestPage.grid.columns.author'),
        name: 'author',
        options: {
          customBodyRender: (author: NamedModel) => (author ? author.name : ''),
        },
      },
      {
        label: t('requestPage.grid.columns.owner'),
        name: 'owner',
        options: {
          display: 'false',
          customBodyRender: (owner: NamedModel) => (owner ? owner.name : ''),
        },
      },
      {
        label: t('requestPage.grid.columns.state'),
        name: 'state',
        options: {
          display: 'false',
          customBodyRender: (state: RequestState) => {
            if (!state) {
              return null;
            }
            switch (state) {
              case RequestState.ACTIVE:
                return <FiberManualRecord style={{ color: 'green', alignSelf: 'center' }} />;
              case RequestState.INACTIVE:
                return <FiberManualRecord style={{ color: 'red' }} />;
              default:
                return null;
            }
          },
        },
      },
      {
        label: t('requestPage.grid.columns.status'),
        name: 'status',
        options: {
          customBodyRender: (status: NamedModel) => {
            if (!status) {
              return null;
            }
            return <RequestStatus status={status} />;
          },
        },
      },
      {
        label: t('requestPage.grid.columns.parameters'),
        name: 'parameterValues',
        options: {
          customBodyRender: (parameterValues: ParamValueShortVer[]) => {
            if (!parameterValues || parameterValues.length === 0) {
              return '';
            }
            let renderParams = parameterValues;
            if (user !== null && user.role !== Role.ADMIN) {
              renderParams = renderParams.filter(
                param => param.modifier !== ParameterModifier.ALLOCATION_PARAMETER
              );
            }
            return parameterShortVersionList(renderParams);
          },
        },
      },
    ];
    return colDefs;
  }

  private handleColumnSortChange(changedColumn: string, direction: SortOrder) {
    const changeParams: Partial<RequestParams> = {
      sortBy: changedColumn as keyof RequestModel,
      order: direction,
    };
    this.handleChange(changeParams);
  }

  private handleResourceTypeSelected = (value: SuggestionItem) => {
    const { resetResourceTypeParameters, setResourceTypeParameters, typeSuggestions } = this.props;
    if (value && value.key) {
      const params = typeSuggestions.find(item => item.id === value.key);
      setResourceTypeParameters(params ? params.parameters : []);
    } else {
      resetResourceTypeParameters();
    }
  };

  private handleChangePage(currentPage: number) {
    const changeParams: Partial<RequestParams> = {
      page: currentPage,
    };
    this.handleChange(changeParams);
  }

  private handleChangeRowsPerPage(numberOfRows: number) {
    const changeParams: Partial<RequestParams> = {
      pageSize: numberOfRows,
    };
    this.handleChange(changeParams);
  }

  private handleFilterChange = (
    value: { [key: string]: FilterValueType },
    clear: boolean | undefined
  ) => {
    if (clear) {
      const filters = Object.assign({}, this.props.filters);
      filters.values = {};
      this.props.updateFilters(filters);
      this.handleChange();
      return;
    }
    const values = this.props.filters.values ? this.props.filters.values : {};
    const { resourceTypeParametersList } = this.props;
    const changeParams: any = {
      typeName: values[FilterField.RESOURCE_TYPE]
        ? values[FilterField.RESOURCE_TYPE] + ''
        : undefined,
      description: values[FilterField.DESCRIPTION]
        ? values[FilterField.DESCRIPTION] + ''
        : undefined,
      statusId: values[FilterField.STATUS] ? values[FilterField.STATUS] + '' : undefined,
    };

    if (resourceTypeParametersList && resourceTypeParametersList.length !== 0) {
      resourceTypeParametersList.forEach(
        (param: FullParameterModel) =>
          (changeParams[param.id] = getFieldValue(param, values[
            FilterField.PARAMS
          ] as GenericFormValue))
      );
    }
    this.handleChange(changeParams);
  };

  private handleChange(changeParams?: any): void {
    const { params } = this.props;
    const newParams: RequestParams = Object.assign(
      {},
      {
        order: params.order,
        page: params.page,
        pageSize: params.pageSize,
      },
      changeParams
    );
    this.props.fetchRequests(
      changeParams
        ? newParams
        : {
            typeName: undefined,
            description: undefined,
            statusId: undefined,
            order: params.order,
            page: params.page,
            pageSize: params.pageSize,
          }
    );
  }

  private handleRedirect(request: RequestModel) {
    const idAsString: string = request.id.toString(10);
    this.props.redirect(Path.REQUEST_CARD.replace(':id', idAsString));
  }
}
