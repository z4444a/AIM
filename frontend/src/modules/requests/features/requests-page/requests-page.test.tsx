import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import { getMockWithTranslationProps } from '../../../../test-utils/with-translation-props-mock';
import { Order } from '../../../resource-pool/Order';
import { Props, RequestsPage } from './requests-page';
import { RequestParams } from '../../../../model/parameters/request-params';

describe('RequestsPage', () => {
  let wrapper: ShallowWrapper;
  beforeEach(() => {
    const withTranslationProps = getMockWithTranslationProps();
    const params: RequestParams = {
      order: Order.ASC,
      pageSize: 10,
      sortBy: 'type',
      page: 0,
      typeName: '',
    };
    const requests = {
      content: [],
      totalElements: 10,
      page: 0,
    };
    const props: Props = {
      params,
      fetchRequests: () => {},
      fetchTypeSuggestions: () => {},
      clearTypeSuggestions: () => {},
      redirect: () => {},
      requests,
      typeSuggestions: [],
      ...withTranslationProps,
    };
    wrapper = shallow(<RequestsPage {...props} />);
  });

  it('should render correctly', () => {
    expect(shallowToJson(wrapper)).toMatchSnapshot();
  });
});
