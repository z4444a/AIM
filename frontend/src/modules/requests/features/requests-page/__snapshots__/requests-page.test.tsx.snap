// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RequestsPage should render correctly 1`] = `
<withI18nextTranslation(WithStyles(AimDataGrid))
  columns={
    Array [
      Object {
        "label": "requestPage.grid.columns.id_i18n",
        "name": "id",
        "options": Object {
          "display": "false",
        },
      },
      Object {
        "label": "requestPage.grid.columns.creation_i18n",
        "name": "creation",
        "options": Object {
          "customBodyRender": [Function],
        },
      },
      Object {
        "label": "requestPage.grid.columns.type_i18n",
        "name": "type",
        "options": Object {
          "customBodyRender": [Function],
        },
      },
      Object {
        "label": "requestPage.grid.columns.project_i18n",
        "name": "project",
        "options": Object {
          "customBodyRender": [Function],
        },
      },
      Object {
        "label": "requestPage.grid.columns.author_i18n",
        "name": "author",
        "options": Object {
          "customBodyRender": [Function],
        },
      },
      Object {
        "label": "requestPage.grid.columns.owner_i18n",
        "name": "owner",
        "options": Object {
          "customBodyRender": [Function],
          "display": "false",
        },
      },
      Object {
        "label": "requestPage.grid.columns.state_i18n",
        "name": "state",
        "options": Object {
          "customBodyRender": [Function],
          "display": "false",
        },
      },
      Object {
        "label": "requestPage.grid.columns.status_i18n",
        "name": "status",
        "options": Object {
          "customBodyRender": [Function],
        },
      },
      Object {
        "label": "requestPage.grid.columns.parameters_i18n",
        "name": "parameterValues",
        "options": Object {
          "customBodyRender": [Function],
        },
      },
    ]
  }
  customToolbar={[Function]}
  data={Array []}
  filterValues={Object {}}
  onChangePage={[Function]}
  onChangeRowsPerPage={[Function]}
  onColumnSortChange={[Function]}
  onFilterChange={[Function]}
  onRowClick={[Function]}
  pageTitle="requestPage.header_i18n"
  renderCustomFilterCmp={[Function]}
  rowsPerPage={10}
  sort={
    Object {
      "direction": "asc",
      "sortBy": "type",
    }
  }
  toggleFilterPanel={[Function]}
  totalPageCount={10}
/>
`;
